File: android/sdk/src/main/java/com/tencent/mtt/hippy/serialization/compatible/Deserializer.java
/* Tencent is pleased to support the open source community by making Hippy available.
 * Copyright (C) 2021 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.mtt.hippy.serialization.compatible;

import com.tencent.mtt.hippy.common.ConstantValue;
import com.tencent.mtt.hippy.common.HippyArray;
import com.tencent.mtt.hippy.common.HippyMap;
import com.tencent.mtt.hippy.exception.UnexpectedException;
import com.tencent.mtt.hippy.exception.UnreachableCodeException;
import com.tencent.mtt.hippy.serialization.ErrorTag;
import com.tencent.mtt.hippy.serialization.PrimitiveValueDeserializer;
import com.tencent.mtt.hippy.serialization.SerializationTag;
import com.tencent.mtt.hippy.serialization.StringLocation;
import com.tencent.mtt.hippy.serialization.exception.DataCloneOutOfRangeException;
import com.tencent.mtt.hippy.serialization.nio.reader.BinaryReader;
import com.tencent.mtt.hippy.serialization.string.StringTable;

import java.math.BigInteger;

/**
 * Compatible with {@code com.tencent.mtt.hippy.common.HippyMap}
 */
@SuppressWarnings({"unused", "deprecation"})
public class Deserializer extends PrimitiveValueDeserializer {

  /**
   * Version 13 compatibility mode
   */
  private boolean version13BrokenDataMode = false;

  public Deserializer(BinaryReader reader) {
    this(reader, null);
  }

  public Deserializer(BinaryReader reader, StringTable stringTable) {
    super(reader, stringTable);
  }

  @Override
  protected int getSupportedVersion() {
    return 15;
  }

  @Override
  protected Object getUndefined() {
    return ConstantValue.Undefined;
  }

  @Override
  protected Object getNull() {
    return ConstantValue.Null;
  }

  @Override
  protected Object getHole() {
    return ConstantValue.Hole;
  }

  @Override
  @SuppressWarnings("fallthrough")
  protected Object readValue(byte tag, StringLocation location, Object relatedKey) {
    Object object = super.readValue(tag, location, relatedKey);
    if (object != Nothing) {
      return object;
    }

    switch (tag) {
      case SerializationTag.TRUE_OBJECT:
        return readJSBoolean(true);
      case SerializationTag.FALSE_OBJECT:
        return readJSBoolean(false);
      case SerializationTag.NUMBER_OBJECT:
        return readJSNumber();
      case SerializationTag.BIG_INT_OBJECT:
        return readJSBigInt();
      case SerializationTag.STRING_OBJECT:
        return readJSString(location, relatedKey);
      case SerializationTag.REGEXP:
        return readJSRegExp();
      case SerializationTag.ARRAY_BUFFER:
        return readJSArrayBuffer();
      case SerializationTag.ARRAY_BUFFER_TRANSFER:
        return readTransferredJSArrayBuffer();
      case SerializationTag.SHARED_ARRAY_BUFFER:
        return readSharedArrayBuffer();
      case SerializationTag.BEGIN_JS_OBJECT:
        return readJSObject();
      case SerializationTag.BEGIN_JS_MAP:
        return readJSMap();
      case SerializationTag.BEGIN_JS_SET:
        return readJSSet();
      case SerializationTag.BEGIN_DENSE_JS_ARRAY:
        return readDenseArray();
      case SerializationTag.BEGIN_SPARSE_JS_ARRAY:
        return readSparseArray();
      case SerializationTag.WASM_MODULE_TRANSFER:
        return readTransferredWasmModule();
      case SerializationTag.HOST_OBJECT:
        return readHostObject();
      case SerializationTag.WASM_MEMORY_TRANSFER:
        return readTransferredWasmMemory();
      case SerializationTag.ERROR:
        return readJSError();
      case SerializationTag.SHARED_OBJECT: {
        if (getWireFormatVersion() >= 15) {
          return readSharedObject();
        }
        // If the data doesn't support shared values because it is from an older
        // version, treat the tag as unknown.
        // [[fallthrough]]
      }
      default: {
        //  Before there was an explicit tag for host objects, all unknown tags
        //  were delegated to the host.
        if (getWireFormatVersion() < 13) {
          reader.position(-1);
          return readHostObject();
        }

        // Unsupported Tag treated as Undefined
        return Undefined;
      }
    }
  }

  /**
   * Deserializes a JavaScript delegate object from the buffer.
   *
   * @return JavaScript delegate object
   */
  @Override
  public Object readValue() {
    /*
     *  The `V8` engine has a bug which produced invalid version 13 data (see https://crbug.com/1284506).
     *
     *  This compatibility mode tries to first read the data normally,
     *  and if it fails, and the version is 13, tries to read the broken format.
     */
    int originalPosition = reader.position();
    try {
      return super.readValue();
    } catch (Exception e) {
      if (getWireFormatVersion() == 13) {
        reader.position(originalPosition);
        version13BrokenDataMode = true;
        return super.readValue();
      } else {
        throw e;
      }
    }
  }

  private ErrorTag readErrorTag() {
    return ErrorTag.fromTag((byte) reader.getVarint());
  }

  private Boolean readJSBoolean(boolean value) {
    return assignId(value);
  }

  private Number readJSNumber() {
    return assignId(reader.getDouble());
  }

  private BigInteger readJSBigInt() {
    return assignId(readBigInt());
  }

  private String readJSString(StringLocation location, Object relatedKey) {
    return assignId(readString(location, relatedKey));
  }

  private Object readJSArrayBuffer() {
    int byteLength = (int) reader.getVarint();
    if (byteLength < 0) {
      throw new DataCloneOutOfRangeException(byteLength);
    }
    reader.position(reader.position() + byteLength);

    assignId(Undefined);
    if (peekTag() == SerializationTag.ARRAY_BUFFER_VIEW) {
      readJSArrayBufferView();
    }

    return null;
  }

  private Object readJSRegExp() {
    readString(StringLocation.VOID, null);
    reader.getVarint();
    return assignId(Undefined);
  }

  private HippyMap readJSObject() {
    HippyMap object = new HippyMap();
    assignId(object);
    int read = readJSProperties(object, SerializationTag.END_JS_OBJECT);
    int expected = (int) reader.getVarint();
    if (read != expected) {
      throw new UnexpectedException("unexpected number of properties");
    }
    return object;
  }

  private int readJSProperties(HippyMap object, byte endTag) {
    final StringLocation keyLocation, valueLocation;
    if (endTag == SerializationTag.END_DENSE_JS_ARRAY) {
      keyLocation = StringLocation.DENSE_ARRAY_KEY;
      valueLocation = StringLocation.DENSE_ARRAY_ITEM;
    } else if (endTag == SerializationTag.END_JS_OBJECT) {
      keyLocation = StringLocation.OBJECT_KEY;
      valueLocation = StringLocation.OBJECT_VALUE;
    } else {
      throw new UnreachableCodeException();
    }

    byte tag;
    int count = 0;
    while ((tag = readTag()) != endTag) {
      count++;
      Object key = readValue(tag, keyLocation, null);
      Object value = readValue(valueLocation, key);

      if (object != null && value != Undefined) {
        if (key instanceof Number) {
          object.pushObject(String.valueOf(key), value);
        } else if (key instanceof String) {
          if (key == "null") {
            object.pushObject(null, value);
          } else {
            object.pushObject((String) key, value);
          }
        } else {
          throw new AssertionError("Object key is not of String(null) nor Number type");
        }
      }
    }
    return count;
  }

  private HippyMap readJSMap() {
    HippyMap object = new HippyMap();
    assignId(object);
    byte tag;
    int read = 0;
    while ((tag = readTag()) != SerializationTag.END_JS_MAP) {
      read++;
      Object key = readValue(tag, StringLocation.MAP_KEY, null);
      key = key.toString();
      Object value = readValue(StringLocation.MAP_VALUE, key);
      if (value != Undefined) {
        if (key == "null") {
          object.pushObject(null, value);
        } else {
          object.pushObject((String) key, value);
        }
      }
    }
    int expected = (int) reader.getVarint();
    if (2 * read != expected) {
      throw new UnexpectedException("unexpected number of entries");
    }
    return object;
  }

  private HippyArray readJSSet() {
    HippyArray array = new HippyArray();
    assignId(array);
    byte tag;
    int read = 0;
    while ((tag = readTag()) != SerializationTag.END_JS_SET) {
      read++;
      Object value = readValue(tag, StringLocation.SET_ITEM, null);
      array.pushObject(value);
    }
    int expected = (int) reader.getVarint();
    if (read != expected) {
      throw new UnexpectedException("unexpected number of values");
    }
    return array;
  }

  private HippyArray readDenseArray() {
    int length = (int) reader.getVarint();
    if (length < 0) {
      throw new DataCloneOutOfRangeException(length);
    }
    HippyArray array = new HippyArray();
    assignId(array);
    for (int i = 0; i < length; i++) {
      byte tag = readTag();
      if (tag != SerializationTag.THE_HOLE) {
        array.pushObject(readValue(tag, StringLocation.DENSE_ARRAY_ITEM, i));
      }
    }

    int read = readJSProperties(null, SerializationTag.END_DENSE_JS_ARRAY);
    int expected = (int) reader.getVarint();
    if (read != expected) {
      throw new UnexpectedException("unexpected number of properties");
    }
    int length2 = (int) reader.getVarint();
    if (length != length2) {
      throw new AssertionError("length ambiguity");
    }
    return array;
  }

  /**
   * Reads Spare Array from buffer.
   *
   * <h2>Note</h2>
   * Sparse arrays will be serialized as an object-like manner. Normally, it should be representable
   * as {@link HippyMap}, but in order to be compatible with the previous serialization implement,
   * we use {@link HippyArray} to express sparse arrays. <br/> When a hole is encountered, {@link
   * ConstantValue#Null} is used to fill it.
   *
   * @return array
   */
  private HippyArray readSparseArray() {
    long length = reader.getVarint();
    HippyArray array = new HippyArray();
    assignId(array);

    byte tag;
    int read = 0;
    while ((tag = readTag()) != SerializationTag.END_SPARSE_JS_ARRAY) {
      read++;
      Object key = readValue(tag, StringLocation.SPARSE_ARRAY_KEY, null);
      Object value = readValue(StringLocation.SPARSE_ARRAY_ITEM, key);

      int index = -1;
      if (key instanceof Number) {
        index = ((Number) key).intValue();
      } else if (key instanceof String) {
        try {
          index = Integer.parseInt((String) key);
        } catch (NumberFormatException ignored) {
          // ignore not parsable string
        }
      }

      if (index >= 0) {
        int spaceNeeded = (index + 1) - array.size();
        if (spaceNeeded
            == 1) { // Fast path, item are ordered in general ECMAScript(VM) implementation
          array.pushObject(value);
        } else {  // Slow path, universal
          for (int i = 0; i < spaceNeeded; i++) {
            array.pushNull();
          }
          array.setObject(index, value);
        }
      }
    }

    int expected = (int) reader.getVarint();
    if (read != expected) {
      throw new UnexpectedException("unexpected number of properties");
    }
    long length2 = reader.getVarint();
    if (length != length2) {
      throw new AssertionError("length ambiguity");
    }
    return array;
  }

  private void readJSArrayBufferView() {
    byte arrayBufferViewTag = readTag();
    if (arrayBufferViewTag != SerializationTag.ARRAY_BUFFER_VIEW) {
      throw new AssertionError("ArrayBufferViewTag: " + arrayBufferViewTag);
    }
    reader.getVarint();
    reader.getVarint();
    if (getWireFormatVersion() >= 14 || version13BrokenDataMode) {
      reader.getVarint();
    }
    reader.getVarint();

    assignId(Undefined);
  }

  private HippyMap readJSError() {
    String message = null;
    String stack = null;
    String errorType = null;
    boolean done = false;
    while (!done) {
      ErrorTag tag = readErrorTag();
      if (tag == null) {
        break;
      }
      switch (tag) {
        case EVAL_ERROR:
          errorType = "EvalError";
          break;
        case RANGE_ERROR:
          errorType = "RangeError";
          break;
        case REFERENCE_ERROR:
          errorType = "ReferenceError";
          break;
        case SYNTAX_ERROR:
          errorType = "SyntaxError";
          break;
        case TYPE_ERROR:
          errorType = "TypeError";
          break;
        case URI_ERROR:
          errorType = "URIError";
          break;
        case MESSAGE:
          message = readString(StringLocation.ERROR_MESSAGE, null);
          break;
        case STACK:
          stack = readString(StringLocation.ERROR_STACK, null);
          break;
        default:
          if (!(tag == ErrorTag.END)) {
            throw new AssertionError("ErrorTag: " + tag);
          }
          done = true;
          break;
      }
    }

    HippyMap error = new HippyMap();
    error.pushString("message", message);
    error.pushString("stack", stack);
    error.pushString("type", errorType);
    assignId(error);
    return error;
  }

  private Object readHostObject() {
    return assignId(Undefined);
  }

  private Object readSharedObject() {
    reader.getVarint();
    return assignId(Undefined);
  }

  private Object readTransferredJSArrayBuffer() {
    reader.getVarint();
    assignId(Undefined);
    if (peekTag() == SerializationTag.ARRAY_BUFFER_VIEW) {
      readJSArrayBufferView();
    }
    return null;
  }

  private Object readSharedArrayBuffer() {
    reader.getVarint();
    assignId(Undefined);
    if (peekTag() == SerializationTag.ARRAY_BUFFER_VIEW) {
      readJSArrayBufferView();
    }
    return null;
  }

  private Object readTransferredWasmModule() {
    reader.getVarint();
    assignId(Undefined);
    return null;
  }

  private Object readTransferredWasmMemory() {
    reader.getVarint();
    readSharedArrayBuffer();
    assignId(Undefined);
    return null;
  }
}

File: renderer/flutter/lib/serialization/deserializer.dart
import '../common/voltron_array.dart';
import '../common/voltron_map.dart';
import 'primitive_value_deserializer.dart';
import 'reader/binary_reader.dart';
import 'serialization_tag.dart';
import 'string/string_location.dart';
import 'string/string_table.dart';

class Deserializer extends PrimitiveValueDeserializer {
  Deserializer(StringTable stringTable) : super(stringTable);

  @override
  Object getHole() {
    return undefined;
  }

  @override
  Object? getNull() {
    return null;
  }

  @override
  Object getUndefined() {
    return Object();
  }

  @override
  VoltronArray readDenseArray() {
    var length = reader.getVarint();
    if (length < 0) {
      throw RangeError("readDenseArray expected length : $length");
    }
    var array = VoltronArray();
    assignId(array);
    for (var i = 0; i < length; i++) {
      var tag = readTag();
      if (tag != SerializationTag.kTheHole) {
        array.push(readValue(
            tag: tag, location: StringLocation.denseArrayItem, relatedKey: i));
      }
    }

    var read = _readJSProperties(null, SerializationTag.kEndDenseJsArray);
    var expected = reader.getVarint();
    if (read != expected) {
      throw UnsupportedError("unexpected number of properties");
    }

    var length2 = reader.getVarint();
    if (length != length2) {
      throw AssertionError("length ambiguity");
    }
    return array;
  }

  @override
  Object readHostObject() {
    return assignId(undefined);
  }

  @override
  Object? readJSArrayBuffer() {
    var byteLength = reader.getVarint();
    if (byteLength < 0) {
      throw RangeError("readJSArrayBuffer expected length: $byteLength");
    }
    reader.position = reader.position + byteLength;
    assignId(undefined);
    if (peekTag() == SerializationTag.kArrayBufferView) {
      readJSArrayBufferView();
    }
    return null;
  }

  @override
  BigInt? readJSBigInt() {
    return assignId(readBigInt());
  }

  @override
  bool readJSBoolean(bool value) {
    return assignId(value);
  }

  @override
  VoltronMap readJSError() {
    var message;
    var stack;
    var errorType;

    var done = false;
    while (!done) {
      var tag = readErrorTag();
      if (tag == ErrorTag.kUnknownTag) {
        break;
      }
      if (tag == ErrorTag.kEvalError) {
        errorType = "EvalError";
        break;
      } else if (tag == ErrorTag.kRangeError) {
        errorType = "RangeError";
        break;
      } else if (tag == ErrorTag.kReferenceError) {
        errorType = "ReferenceError";
        break;
      } else if (tag == ErrorTag.kSyntaxError) {
        errorType = "SyntaxError";
        break;
      } else if (tag == ErrorTag.kTypeError) {
        errorType = "TypeError";
        break;
      } else if (tag == ErrorTag.kURIError) {
        errorType = "URIError";
        break;
      } else if (tag == ErrorTag.kMessage) {
        message = readString(StringLocation.errorMessage, null);
        break;
      } else if (tag == ErrorTag.kStack) {
        stack = readString(StringLocation.errorStack, null);
        break;
      } else {
        if (tag != ErrorTag.kEnd) {
          throw AssertionError("ErrorTag: $tag");
        }
        done = true;
        break;
      }
    }

    var error = VoltronMap();
    error.push("message", message);
    error.push("stack", stack);
    error.push("type", errorType);
    assignId(error);
    return error;
  }

  @override
  VoltronMap readJSMap() {
    var object = VoltronMap();
    assignId(object);
    int tag;
    var read = 0;
    while ((tag = readTag()) != SerializationTag.kEndJsMap) {
      read++;
      var key = readValue(tag: tag, location: StringLocation.mapKey).toString();
      var value = readValue(location: StringLocation.mapValue, relatedKey: key);
      if (value != undefined) {
        if (key == "null") {
          object.push("NULL", value);
        } else {
          object.push(key, value);
        }
      }
    }
    var expected = reader.getVarint();
    if (2 * read != expected) {
      throw UnsupportedError("unexpected number of entries");
    }
    return object;
  }

  @override
  num readJSNumber() {
    return assignId(reader.getDouble());
  }

  @override
  VoltronMap readJSObject() {
    var object = VoltronMap();
    assignId(object);
    var read = _readJSProperties(object, SerializationTag.kEndJsObject);
    var expected = reader.getVarint();
    if (read != expected) {
      throw UnsupportedError("unexpected number of properties");
    }
    return object;
  }

  @override
  Object? readJSRegExp() {
    readString(StringLocation.eVoid, null);
    reader.getVarint();
    return assignId(undefined);
  }

  @override
  VoltronArray readJSSet() {
    var array = VoltronArray();
    assignId(array);
    int tag;
    var read = 0;
    while ((tag = readTag()) != SerializationTag.kEndJsSet) {
      read++;
      var value = readValue(tag: tag, location: StringLocation.setItem);
      array.push(value);
    }
    var expected = reader.getVarint();
    if (read != expected) {
      throw UnsupportedError("unexpected number of values");
    }
    return array;
  }

  @override
  String? readJSString(StringLocation location, Object? relatedKey) {
    return assignId(readString(location, relatedKey));
  }

  @override
  Object? readSharedArrayBuffer() {
    reader.getVarint();
    assignId(undefined);
    if (peekTag() == SerializationTag.kArrayBufferView) {
      readJSArrayBufferView();
    }
    return null;
  }

  @override
  VoltronArray readSparseArray() {
    var length = reader.getVarint();
    var array = VoltronArray();
    assignId(array);

    int tag;
    var read = 0;
    while ((tag = readTag()) != SerializationTag.kEndSparseJsArray) {
      read++;
      var key = readValue(tag: tag, location: StringLocation.sparseArrayKey);
      var value =
          readValue(location: StringLocation.sparseArrayItem, relatedKey: key);

      var index = -1;
      if (key is num) {
        index = key.toInt();
      } else if (key is String) {
        try {
          index = int.parse(key);
        } on FormatException {
          // ignore not parsable string
        }
      }
      if (index >= 0) {
        var spaceNeeded = (index + 1) - array.size();
        if (spaceNeeded == 1) {
          // Fast path, item are ordered in general ECMAScript(VM) implementation
          array.push(value);
        } else {
          // Slow path, universal
          for (var i = 0; i < spaceNeeded; i++) {
            array.pushNull();
          }
          array.set(index, value);
        }
      }
    }

    var expected = reader.getVarint();
    if (read != expected) {
      throw UnsupportedError("unexpected number of properties");
    }
    var length2 = reader.getVarint();
    if (length != length2) {
      throw AssertionError("length ambiguity");
    }
    return array;
  }

  @override
  Object? readTransferredJSArrayBuffer() {
    reader.getVarint();
    assignId(undefined);
    if (peekTag() == SerializationTag.kArrayBufferView) {
      readJSArrayBufferView();
    }
    return null;
  }

  @override
  Object? readTransferredWasmMemory() {
    reader.getVarint();
    readSharedArrayBuffer();
    assignId(undefined);
    return null;
  }

  @override
  Object? readTransferredWasmModule() {
    reader.getVarint();
    assignId(undefined);
    return null;
  }

  int _readJSProperties(VoltronMap? object, int endTag) {
    final StringLocation keyLocation, valueLocation;
    if (endTag == SerializationTag.kEndDenseJsArray) {
      keyLocation = StringLocation.denseArrayKey;
      valueLocation = StringLocation.denseArrayItem;
    } else if (endTag == SerializationTag.kEndJsObject) {
      keyLocation = StringLocation.objectKey;
      valueLocation = StringLocation.objectValue;
    } else {
      throw ArgumentError("_readJSProperties Illegal endTag: $endTag");
    }

    int tag;
    var count = 0;
    while ((tag = readTag()) != endTag) {
      count++;
      var key = readValue(tag: tag, location: keyLocation);
      var value = readValue(location: valueLocation, relatedKey: key);

      if (object != null && value != undefined) {
        if (key is num) {
          object.push(key.toString(), value);
        } else if (key is String) {
          if (key == "null") {
            object.push("NULL", value);
          } else {
            object.push(key.toString(), value);
          }
        } else {
          throw AssertionError(
              "Object key is not of String(null) nor Number type");
        }
      }
    }
    return count;
  }

  void readJSArrayBufferView() {
    var arrayBufferViewTag = readTag();
    if (arrayBufferViewTag != SerializationTag.kArrayBufferView) {
      throw AssertionError(
          "readJSArrayBufferView ArrayBufferViewTag: $arrayBufferViewTag");
    }

    reader.getVarint();
    reader.getVarint();
    readArrayBufferViewTag();

    assignId(undefined);
  }
}
